File: src/jl/ppls/ppl.jl

abstract type PPL end

struct VariableDefinition
    node::SyntaxNode
end

struct Model
    name::Symbol
    node::SyntaxNode
end

function is_model(::PPL, node::SyntaxNode)::Bool
    error("Not implemented.")
end

function get_model_name(::PPL, node::SyntaxNode)::Symbol
    error("Not implemented.")
end

function is_random_variable_definition(::PPL, node::SyntaxNode)::Bool
    error("Not implemented.")
end

function get_random_variable_name(ppl::PPL, variable::VariableDefinition)::Symbol
    error("Not implemented.")
end

function get_address_node(ppl::PPL, variable::VariableDefinition)::SyntaxNode
    error("Not implemented.")
end

function is_observed(ppl::PPL, variable::VariableDefinition)::Bool
    error("Not implemented.")
end

function get_distribution_node(ppl::PPL, variable::VariableDefinition)::SyntaxNode
    error("Not implemented.")
end

# parameters follow convention of Wikipedia
# returns distribution name + map of paramter name to syntax node
function get_distribution(ppl::PPL, distribution_node::SyntaxNode)::Tuple{String, Dict{String,SyntaxNode}}
    error("Not implemented.")
end

function preprocess_syntaxtree!(ppl::PPL, syntax_tree::SyntaxTree)
    error("Not implemented.")
end

File: src/py/ppls/ppl.py
import ast
from typing import Union
from ast_utils.preprocess import SyntaxTree

class VariableDefinition:
    def __init__(self, node: ast.AST):
        self.node = node

class Model:
    def __init__(self, name: str, node: ast.AST):
        self.name = name
        self.node = node

class PPL:
    def __init__(self) -> None:
        pass

    def is_random_variable_definition(self, node: ast.AST) -> bool:
        raise NotImplementedError
    
    def is_model(self, node: ast.AST) -> bool:
        raise NotImplementedError
    
    def get_model_name(self, node: ast.AST) -> str:
        raise NotImplementedError
    
    def get_random_variable_name(self, variable: VariableDefinition) -> str:
        raise NotImplementedError
    
    def get_address_node(self, variable: VariableDefinition) -> ast.AST:
        raise NotImplementedError
    
    def is_observed(self, variable: VariableDefinition) -> bool:
        raise NotImplementedError
    
    def get_distribution_node(self, variable: VariableDefinition) -> ast.AST:
        raise NotImplementedError
    
    def get_distribution(self, distribution_node: ast.AST) -> tuple[str, dict[str, ast.AST]]:
        raise NotImplementedError
    
    def get_program_variable_name(self, variable: VariableDefinition) -> Union[str,None]:
        raise NotImplementedError

    def preprocess_syntax_tree(self, syntax_tree: SyntaxTree) -> SyntaxTree:
        raise NotImplementedError